datafusion/optimizer/src/replace_distinct_aggregate.rs

```diff
@@ -99,17 +97,7 @@ impl OptimizerRule for ReplaceDistinctWithAggregate {
                 // Construct the aggregation expression to be used to fetch the selected expressions.
                 let aggr_expr = select_expr
                     .into_iter()
-                    .map(|e| {
-                        Expr::AggregateFunction(AggregateFunction::new(
-                            AggregateFunctionFunc::FirstValue,
-                            vec![e],
-                            false,
-                            None,
-                            sort_expr.clone(),
-                            None,
-                        ))
-                    })
-                    .collect::<Vec<Expr>>();
+                    .map(|e| first_value(vec![e], false, None, sort_expr.clone(), None));
```

Comment:
I just noticed that we import functions-aggregate in optimizer for first_value, should we move this rule into functions-aggregate?

Reply:
As far as I know, All `LogicalPlan` rules are under `optimizer/src/` folder. Also, I don't see any downside in having this rule here. Hence, I think it is better to keep rule in its current place. I wonder what others think though?

Reply:
Perhaps what @jayzhan211  is noticing is that by directly relying on `datafusion-aggregate-functions` in the optimizer it means that now the optimizer is still treating "built in" aggregates specially (and for example, user defined functions couldn't take advantage of this)

So if we moved this rule (or somehow made the dependency via an interface) it would be more general and decouple the code and make datafusion more extensible. 

One way is to move the entire rule into `datafusion-aggregate-functions` but maybe we could instead have `replace_distinct_aggregate` depend on `FunctionRegistry` rather than a direct call to `first_value` for example

Maybe we can file a TODO ticket to track it